File: python/aberp-cad-extract/aberp_cad_extract/feature_graph.py
"""FeatureGraph wire schema — mirrors ``aberp_quote_engine::FeatureGraph``.

The contract: a Python-produced FeatureGraph JSON MUST deserialize
cleanly into the Rust struct without data loss. Field names, types,
and enum string-forms are pinned by the Rust side; this module is the
Python shadow of that pin.

Per [[aberp-quoting-design-addenda]] addendum 1, both
``requires_5_axis`` and ``thin_wall_present`` are first-class
booleans, not derived counters. They are required on every output —
``Optional[bool]`` is intentionally not used. Missing or wrong-type
fails Pydantic validation and the cross-language compat test in the
Rust crate.
"""

from __future__ import annotations

from enum import Enum
from typing import List

from pydantic import BaseModel, ConfigDict, Field


#: Wire schema version. Must land inside the Rust wrapper's accepted
#: range ``MIN_SCHEMA_VERSION..=EXPECTED_SCHEMA_VERSION`` (2..=6), where
#: the ceiling is ``aberp_quote_engine::FeatureGraph::SCHEMA_VERSION``.
#:
#: **ADR-0112 B.1 note.** Before that cut the wrapper's guard was exact
#: equality against 2, so bumping this constant without editing the Rust
#: side in the same diff would have failed EVERY extraction Permanent.
#: The guard is now a range, which is what makes this 2 -> 6 bump safe:
#: it lands inside the range, and a stored graph stamped at any older
#: version keeps loading with the newer fields defaulted.
#:
#: Jumped 2 -> 6, not 2 -> 3: v3/v4/v5 fields (``stock_form``, ``gears``,
#: ``tolerance``) already exist on the Rust struct and are supplied by
#: the operator paths, not by this extractor. Catching the extractor up
#: to them is a SEPARATE workstream (ADR-0112 Part B0) because it moves
#: prices; v6 does not.
#:
#: **This bump CHANGES ``feature_graph_hash`` for every freshly extracted
#: part — including parts with no holes.** Stated plainly here because an
#: earlier revision of these docs claimed the opposite. The daemon hashes
#: ``blake3(serde_json::to_vec(&graph))``
#: (``apps/aberp/src/quote_pricing_pipeline.rs``), that encoding contains
#: ``_schema_version``, and this constant is inside it. Omitting an empty
#: ``located_holes`` keeps the REST of the encoding identical, which is
#: worth doing and is pinned — but it does not and cannot hold the
#: version field still.
#:
#: Why the jump is nevertheless the right call rather than something to
#: minimise:
#:
#: * The number has to move at all, because v6 is the version at which
#:   ``located_holes`` may be present and the wrapper's accepted range is
#:   what tells the two sides apart. Stamping 2 while emitting v6 content
#:   would be a lie in the field whose entire job is to describe the
#:   content.
#: * Stamping 3, 4 or 5 instead would be a DIFFERENT lie — those versions
#:   mean "carries ``stock_form`` / ``gears`` / ``tolerance``", which this
#:   extractor does not produce. There is no smaller honest step; the
#:   version line is a ladder, not a bitmask.
#: * The blast radius is bounded and was checked, not assumed: nothing in
#:   the tree COMPARES a ``feature_graph_hash`` against a stored one. It
#:   is written into the ``quote.pricing_extracted`` audit payload and
#:   posted to the storefront as a content tag; no cache key, no
#:   idempotency short-circuit, no golden asserts a literal value. So the
#:   effect is that parts re-extracted after this deploy get a new tag —
#:   once — and no price, no PDF and no stored breakdown moves.
#:
#: What IS byte-identical, and is pinned rather than asserted in prose:
#: a stored ≤v5 graph re-serialises to exactly its old bytes (it keeps
#: its own stamped version), and the frozen Portable edition never runs
#: this extractor at all, so its encoding cannot move. See
#: ``crates/aberp-quote-engine/tests/feature_graph_compat.rs``.
SCHEMA_VERSION: int = 6


class FeatureType(str, Enum):
    """Closed-vocab feature kinds.

    Strings match the Rust ``FeatureType`` serde rename (snake_case)
    AND the ``quoting_complexity_rules.feature_type`` DB column (S267).
    """

    POCKET = "pocket"
    HOLE = "hole"
    SLOT = "slot"
    THREAD = "thread"
    UNDERCUT_5_AXIS = "undercut_5axis"
    THIN_WALL = "thin_wall"
    SURFACE = "surface"
    ENGRAVING = "engraving"


class Feature(BaseModel):
    """One feature on the extracted part.

    Mirrors the Rust ``Feature`` struct field-for-field.
    """

    model_config = ConfigDict(extra="forbid")

    feature_type: FeatureType
    count: int = Field(ge=1)
    representative_size_mm: float = Field(ge=0.0)


class HoleEndCondition(str, Enum):
    """How a located hole terminates along its axis.

    Mirrors the Rust ``HoleEndCondition`` serde strings (snake_case).

    ``UNKNOWN`` is FIRST-CLASS, never a silent default to ``THROUGH``: a
    blind hole mis-read as through under-counts peck cycles and therefore
    under-prices. An extractor that cannot tell must say so.
    """

    THROUGH = "through"
    BLIND = "blind"
    UNKNOWN = "unknown"


class LocatedHole(BaseModel):
    """One cylindrical hole with a known axis, depth, diameter, position.

    **ADR-0112 Part B (schema v6).** Mirrors the Rust ``LocatedHole``
    struct field-for-field. Millimetres, in the part's own coordinate
    system (the STEP reader normalises any declared ``LENGTH_UNIT`` to MM
    on import). No WCS, no fixture offset.

    ``axis_unit`` is normalised HERE, by the extractor — the engine "does
    not second-guess the extractor". It points from ``entry_point_mm``
    into the material.
    """

    model_config = ConfigDict(extra="forbid")

    diameter_mm: float = Field(gt=0.0)
    depth_mm: float = Field(gt=0.0)
    axis_unit: List[float] = Field(min_length=3, max_length=3)
    entry_point_mm: List[float] = Field(min_length=3, max_length=3)
    end_condition: HoleEndCondition = HoleEndCondition.UNKNOWN
    flat_bottom: bool = False


class FeatureGraph(BaseModel):
    """Extracted-geometry side of the quote-engine input.

    Mirrors the Rust ``FeatureGraph`` struct. The Rust side renames
    the schema-version field to ``_schema_version``; Pydantic alias
    + ``populate_by_name`` lets us round-trip with both names.

    Addendum 1 booleans are required fields: omitting or nulling
    either is a validation error, not a runtime fallback.
    """

    model_config = ConfigDict(
        extra="forbid",
        populate_by_name=True,
    )

    schema_version: int = Field(
        default=SCHEMA_VERSION,
        alias="_schema_version",
    )
    bounding_box_mm: List[float] = Field(min_length=3, max_length=3)
    volume_mm3: float = Field(ge=0.0)
    # Schema v2 (S418): total surface area in mm², drives finishing-pass
    # machining time in the engine (report §5.2). STL = Σ ½‖(v1−v0)×
    # (v2−v0)‖ over triangles; STEP = ``BRepGProp.SurfaceProperties``.
    surface_area_mm2: float = Field(ge=0.0)
    material_grade: str = Field(min_length=1)
    features: List[Feature]

    requires_5_axis: bool
    thin_wall_present: bool

    #: **ADR-0112 Part B (v6).** Located holes mined from the STEP B-rep.
    #: Defaults to empty, and ``to_canonical_dict`` OMITS the key when it
    #: is empty (see there), so a hole-less part adds no bytes here.
    #:
    #: CORRECTED (ADR-0112 adversarial S1): that does NOT make a freshly
    #: extracted hole-less part's ``feature_graph_hash`` unchanged, which
    #: is what this comment used to claim. ``_schema_version`` moved 2 ->
    #: 6 in the same cut and is inside the hashed encoding, so the hash
    #: moves for every extracted part regardless of this field. See
    #: :data:`SCHEMA_VERSION` for the full accounting and for what byte
    #: identity genuinely does still hold.
    located_holes: List[LocatedHole] = Field(default_factory=list)

    def to_canonical_dict(self) -> dict:
        """Plain-JSON dict using the wire field names (Rust-compatible).

        The Rust deserializer reads ``_schema_version``; this is the
        single point that turns the Python field name into the wire
        name before ``json.dumps`` is called by the CLI.

        **ADR-0112 v6.** An EMPTY ``located_holes`` emits no key at all,
        mirroring the Rust side's ``skip_serializing_if = "Vec::is_empty"``.
        Both halves of the wire contract have to agree on this, because
        the daemon blake3-hashes this exact encoding into
        ``feature_graph_hash`` and a graph that round-trips through Rust
        must come back as the same bytes it went in as.

        Scoped honestly (ADR-0112 adversarial S1): this keeps the two
        LANGUAGES' encodings identical to each other. It does not keep a
        freshly extracted part's hash identical to its pre-v6 value —
        ``_schema_version`` moved 2 -> 6 in the same cut and is inside the
        same hashed encoding. See :data:`SCHEMA_VERSION`.
        """
        out = self.model_dump(by_alias=True, mode="json")
        if not out.get("located_holes"):
            out.pop("located_holes", None)
        return out

File: crates/aberp-quote-engine/src/feature_graph.rs
//! The geometry-side input contract — what the Python extractor (S269)
//! produces and the engine consumes.
//!
//! **Stub for v1.** S269 brings the canonical Python-produced JSON
//! schema; this struct mirrors the agreed shape. Don't expand it
//! without coordinating with S269 — the schema is the wire contract
//! between the two crates. Versioned via [`FeatureGraph::SCHEMA_VERSION`].

use serde::{Deserialize, Serialize};

/// One feature on the part — pocket, hole, slot, …
///
/// Closed-vocab; mirrors [`apps/aberp/src/quoting_tunables.rs`'s
/// `FeatureType` enum (S267)] verbatim so a stored
/// `quoting_complexity_rules.feature_type` string round-trips through
/// the engine without translation. The serde rename to lowercase +
/// `snake_case` matches the DB storage strings.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FeatureType {
    /// Closed cavity / boss pocket.
    Pocket,
    /// Drilled or bored hole.
    Hole,
    /// Linear or radial slot.
    Slot,
    /// Threaded hole or external thread.
    Thread,
    /// 5-axis-only undercut (B-axis or C-axis access).
    ///
    /// Explicit rename: serde's `snake_case` rule breaks the camel-
    /// case around the digit and produces `undercut5_axis`, which
    /// disagrees with the DB string `undercut_5axis` (see
    /// [`FeatureType::as_db_str`]) AND with the S269 Python
    /// extractor's emitted JSON. Aligned here; surfaced by the
    /// S269 cross-language compat test.
    #[serde(rename = "undercut_5axis")]
    Undercut5Axis,
    /// Wall thinner than the operator-tunable thin-wall threshold.
    ThinWall,
    /// Finishing surface that needs its own toolpath.
    Surface,
    /// Engraved / etched feature.
    Engraving,
}

impl FeatureType {
    /// The DB storage string used by `quoting_complexity_rules.feature_type`
    /// (S267). Kept in sync with the `apps/aberp` enum's `as_db_str`.
    pub fn as_db_str(self) -> &'static str {
        match self {
            Self::Pocket => "pocket",
            Self::Hole => "hole",
            Self::Slot => "slot",
            Self::Thread => "thread",
            Self::Undercut5Axis => "undercut_5axis",
            Self::ThinWall => "thin_wall",
            Self::Surface => "surface",
            Self::Engraving => "engraving",
        }
    }
}

/// The size-bucket axis. Mirrors the S267 enum verbatim, including
/// the v1 boundaries returned by [`SizeBucket::range_mm`] — engine
/// and tunables MUST agree on the buckets or a feature could bucket
/// into a bucket no rule exists for. Kept duplicated here (not via
/// dependency on `apps/aberp`) so the engine stays application-
/// independent per design doc §2.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE")]
pub enum SizeBucket {
    /// 0 ≤ size < 10 mm.
    Xs,
    /// 10 ≤ size < 30 mm.
    S,
    /// 30 ≤ size < 80 mm.
    M,
    /// 80 ≤ size < 200 mm.
    L,
    /// 200 mm ≤ size (open-ended).
    Xl,
}

impl SizeBucket {
    /// DB storage string — matches `quoting_complexity_rules.size_bucket`.
    pub fn as_db_str(self) -> &'static str {
        match self {
            Self::Xs => "XS",
            Self::S => "S",
            Self::M => "M",
            Self::L => "L",
            Self::Xl => "XL",
        }
    }

    /// Half-open millimetre range `[min, max)`. `None` upper bound is
    /// open-ended (XL ≥ 200mm). Identical to S267's `range_mm`.
    pub fn range_mm(self) -> (f64, Option<f64>) {
        match self {
            Self::Xs => (0.0, Some(10.0)),
            Self::S => (10.0, Some(30.0)),
            Self::M => (30.0, Some(80.0)),
            Self::L => (80.0, Some(200.0)),
            Self::Xl => (200.0, None),
        }
    }

    /// Bucket a representative size (mm) into the appropriate bucket.
    /// Negative input is clamped to [`SizeBucket::Xs`] (the extractor
    /// should never produce negatives; defence-in-depth).
    pub fn bucket(size_mm: f64) -> Self {
        // Iterate buckets in ascending order; first whose upper bound
        // exceeds the input wins.
        for b in [Self::Xs, Self::S, Self::M, Self::L] {
            if let (_, Some(upper)) = b.range_mm() {
                if size_mm < upper {
                    return b;
                }
            }
        }
        Self::Xl
    }
}

/// Target tolerance class for the part.
///
/// Mirrors S267's `ToleranceRange`. Ordered Loose < Standard < Tight
/// < Precision < UltraPrecision — derived `PartialOrd` works because
/// the variants are declared in that order.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ToleranceRange {
    /// Loosest band — fewer inspection passes, lower multiplier.
    Loose,
    /// Default — design doc baseline.
    Standard,
    /// Tighter tolerance; raises multiplier and inspection time.
    Tight,
    /// Precision band — multi-pass inspection, slower toolpath.
    Precision,
    /// Highest band — CMM time per feature, thermal-stable runs.
    UltraPrecision,
}

impl ToleranceRange {
    /// DB storage string — matches `quoting_tolerance_multipliers.tolerance_range`.
    pub fn as_db_str(self) -> &'static str {
        match self {
            Self::Loose => "loose",
            Self::Standard => "standard",
            Self::Tight => "tight",
            Self::Precision => "precision",
            Self::UltraPrecision => "ultra_precision",
        }
    }
}

/// One feature on the extracted part. The engine processes features
/// in input order — the extractor is responsible for any normalising
/// sort it wants applied.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Feature {
    /// What kind of feature it is.
    pub feature_type: FeatureType,
    /// How many of this feature appear on the part. Engine
    /// distributes the time as `base_time × count × multiplier`.
    pub count: u32,
    /// A characteristic length (mm) for the feature — bore diameter,
    /// pocket width, slot length, etc. Drives the [`SizeBucket`]
    /// lookup. The extractor picks the dimension; the engine does
    /// not second-guess.
    pub representative_size_mm: f64,
}

/// The raw-stock envelope the part is cut from — the shape the shop
/// actually buys and bills, and the volume roughing removes down to the
/// finished part (ADR-0094 Gap 1).
///
/// **Back-compat (load-bearing).** `#[serde(default)]` on
/// [`FeatureGraph::stock_form`] means a graph that omits the field —
/// every persisted blob and every v2-extractor output — loads as
/// [`StockForm::RectangularBlock`], whose volume is exactly today's
/// bounding-box block (`bx·by·bz`). So every existing golden,
/// determinism and property number stays byte-identical until the
/// extractor/operator supplies a round form (Gap 1 part B, S2). Mirrors
/// the `surface_area_mm2` (S418) / `calibration_coefficient` (S429)
/// serde-default precedent.
///
/// Each variant carries its own dimensions: the pure engine evaluates a
/// volume formula and never infers a spin axis from the bounding box —
/// axis inference is the extractor's job, consistent with
/// [`Feature::representative_size_mm`] ("the engine does not
/// second-guess").
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize, Default)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum StockForm {
    /// A rectangular block sized to the bounding box — **today's
    /// model**, reproduced byte-for-byte. Billed/roughed volume is
    /// `bounding_box_mm[0] * [1] * [2]`. The default.
    #[default]
    RectangularBlock,
    /// A solid round bar (turned stock). Billed/roughed volume is the
    /// cylinder `π/4 · diameter_mm² · length_mm` — `0.7854×` the
    /// bounding-box block, so a turned part bills ~21.5 % less material
    /// (and roughs only what a near-net bar actually removes).
    RoundBar {
        /// Bar diameter in millimetres.
        diameter_mm: f64,
        /// Bar cut-off length in millimetres.
        length_mm: f64,
    },
    /// A hollow tube / ring blank. Billed/roughed volume is the annulus
    /// `π/4 · (od_mm² − id_mm²) · length_mm`. The bore is **never
    /// bought**, so it is neither billed as material nor "roughed away"
    /// — the correct model for a ring-gear blank.
    Tube {
        /// Outside diameter in millimetres.
        od_mm: f64,
        /// Inside (bore) diameter in millimetres.
        id_mm: f64,
        /// Tube length in millimetres.
        length_mm: f64,
    },
}

/// **ADR-0094 Gap 3.** The kind of gear teeth to generate. Sets the
/// admissible process family: external spur/helical teeth are hobbed or
/// (in-cycle on a turn-mill) power-skived; internal ring teeth have no
/// hob/skive access and are shaped, broached, or wire-EDM'd.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GearKind {
    /// External spur or helical teeth on the outside of the blank.
    ExternalSpurHelical,
    /// Internal ring-gear teeth (cut from the bore).
    InternalRing,
}

impl GearKind {
    /// DB / wire storage string — matches the future `quoting_gear_ops.kind`.
    pub fn as_db_str(self) -> &'static str {
        match self {
            Self::ExternalSpurHelical => "external_spur_helical",
            Self::InternalRing => "internal_ring",
        }
    }
}

/// **ADR-0094 Gap 3.** The tooth-generation process. [`GearProcess::Auto`]
/// lets the engine pick deterministically from `kind` + routed machine
/// family + AGMA quality (see [`crate::select_gear_process`]); the explicit
/// variants force a process (operator override, reasoning-logged).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GearProcess {
    /// Engine selects the process deterministically.
    Auto,
    /// Hobbing — external teeth on a dedicated hobber (standalone op).
    Hob,
    /// Power-skiving — external teeth in-cycle on a turn-mill (the part is
    /// already on the spindle: no second op, no refixture ⇒ cheap).
    PowerSkive,
    /// Gear shaping — internal ring teeth (and external where blocked).
    Shape,
    /// Broaching — internal teeth: high setup, fast per part at volume.
    Broach,
    /// Wire-EDM — internal teeth at the tightest AGMA classes; slow + dear.
    WireEdm,
}

impl GearProcess {
    /// DB / wire storage string — matches [`crate::GearProcessRate::process`].
    pub fn as_db_str(self) -> &'static str {
        match self {
            Self::Auto => "auto",
            Self::Hob => "hob",
            Self::PowerSkive => "power_skive",
            Self::Shape => "shape",
            Self::Broach => "broach",
            Self::WireEdm => "wire_edm",
        }
    }
}

/// **ADR-0094 Gap 3.** One gear's teeth to cut, modelled as a costed
/// operation — the volume/feature model cannot see teeth. Carries the
/// parameters that drive generation cost (module, tooth count, face width,
/// AGMA quality) and the process selection. Defaulted-empty on
/// [`FeatureGraph::gears`], so a part with no gears prices exactly as today.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GearOp {
    /// External spur/helical vs internal ring — sets the admissible process.
    pub kind: GearKind,
    /// Module `m` (mm) — the tooth-size unit. Bigger module ⇒ more metal per
    /// tooth ⇒ slower generation (time scales by `module_mm^module_exponent`).
    pub module_mm: f64,
    /// Tooth count `z`. Generation time scales linearly with teeth.
    pub teeth: u32,
    /// Face width `b` (mm) — the axial length of the teeth.
    pub face_width_mm: f64,
    /// Target AGMA quality class (higher = tighter = slower). Drives the
    /// quality factor and, for internal rings under `Auto`, the
    /// shape→wire-EDM escalation above [`crate::GEAR_INTERNAL_WIRE_EDM_AGMA`].
    pub quality_agma: u8,
    /// The process to use; [`GearProcess::Auto`] ⇒ engine selects.
    pub process: GearProcess,
}

/// The extracted-geometry side of the engine input.
///
/// Produced by `aberp-cad-extract` (Python, S269) and validated +
/// version-stamped by `aberp-cad-extract-wrapper` (Rust, S270).
/// Plain serde — no clock, no I/O.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FeatureGraph {
    /// `_schema_version` lets S269 evolve the schema without breaking
    /// the wire. The wrapper refuses graphs with an unknown version.
    #[serde(rename = "_schema_version")]
    pub schema_version: u32,
    /// XYZ bounding-box extent in millimetres. S418: now a first-class
    /// scoring input — material is billed on the stock block
    /// `bbox × (1 + scrap_factor)` (report §6.4) and roughing time is
    /// driven by `stock − part` removed volume (report §5.1).
    pub bounding_box_mm: [f64; 3],
    /// Solid volume in mm³ (after extraction, before scrap).
    pub volume_mm3: f64,
    /// Total surface area in mm² (schema v2, S418). Drives the
    /// finishing-pass machining time (report §5.2). STL: Σ ½‖(v1−v0)×
    /// (v2−v0)‖ over triangles; STEP: `BRepGProp::SurfaceProperties`.
    /// `#[serde(default)]` keeps the engine fail-soft on a v1 graph or
    /// a corrupt extractor: a non-positive value falls back to the
    /// bounding-box surface area `2(xy+yz+zx)` inside the engine
    /// (report §5.4) rather than zeroing finishing time.
    #[serde(default)]
    pub surface_area_mm2: f64,
    /// Material grade as it appears in `quoting_materials.grade`
    /// (S266) — e.g. `6061-T6`. The engine errors with
    /// [`crate::QuoteError::MaterialNotInCatalogue`] if no row matches.
    pub material_grade: String,
    /// The features themselves. Engine processes in input order.
    pub features: Vec<Feature>,
    /// **Addendum 1 (per [[aberp-quoting-design-addenda]]).** First-
    /// class boolean for the 5-axis routing decision. The extractor
    /// will set this when any feature requires a 5-axis machine
    /// (compound angle, undercut, 3+2 setup that can't be split).
    /// Engine reads it to set [`crate::QuoteBreakdown::route_to_5_axis`].
    ///
    /// Wired now per the addendum's "land first-class handling
    /// already" instruction; the value will be populated by S269.
    pub requires_5_axis: bool,
    /// **Addendum 1.** First-class boolean for thin-wall presence.
    /// Drives the tight-tolerance labor bump
    /// ([`crate::THIN_WALL_TIGHT_TOL_BUMP`]).
    pub thin_wall_present: bool,
    /// **ADR-0094 Gap 1.** The raw-stock envelope this part is cut from.
    /// Drives BOTH the billed material volume AND the roughing
    /// removed-volume `(stock − part)`. `#[serde(default)]` (mirroring
    /// `surface_area_mm2` / `calibration_coefficient`) makes a graph that
    /// omits it load as [`StockForm::RectangularBlock`] = today's
    /// bounding-box block, so existing goldens stay byte-identical. The
    /// extractor (S269) / operator (S2 wiring) populates it for turned
    /// and tubular parts.
    #[serde(default)]
    pub stock_form: StockForm,
    /// **ADR-0094 Gap 3.** Per-gear tooth-generation operations. Defaulted
    /// empty (`#[serde(default)]`, mirroring `stock_form` / `surface_area_mm2`)
    /// so a graph that omits it — every persisted blob, every pre-S5 extractor
    /// output — prices with NO gear cost and byte-identical output. The
    /// extractor (S269) / operator (S6 wiring) populates it for geared parts.
    #[serde(default)]
    pub gears: Vec<GearOp>,
}

impl FeatureGraph {
    /// The schema version this build of the engine understands. The
    /// wrapper (S270) compares this against the value in the JSON
    /// and refuses unknown versions loud. **v2 (S418)** added
    /// `surface_area_mm2`; the Python `SCHEMA_VERSION` and the
    /// wrapper's `EXPECTED_SCHEMA_VERSION` bump in lockstep. **v3
    /// (ADR-0094 Gap 1)** adds the defaulted `stock_form`; a v2 graph
    /// (no `stock_form`) still loads — it defaults to `RectangularBlock`
    /// — and the version guard accepts any `schema_version ≤ 3`. The
    /// extractor's lockstep bump to v3 lands with S269 (ADR-0094 Q3).
    /// **v4 (ADR-0094 Gap 3)** adds the defaulted `gears` vector; a v2/v3
    /// graph (no `gears`) still loads — it defaults to empty ⇒ zero gear cost
    /// ⇒ today's price — and the guard accepts any `schema_version <= 4`. The
    /// extractor's lockstep bump to v4 lands with S269 (ADR-0094 Q3).
    pub const SCHEMA_VERSION: u32 = 4;
}
